File: src/server/cleaner.ts
/**
 * PulseOps - Data Validation and Cleaning Engine
 * Strict schema verification, data integrity audits, and rejection tracking.
 */
export interface RawInputRecord {
  source_id: string;
  period_from: string;
  period_to: string;
  entity_id: string;
  entity_name: string;
  region: string;
  metric_name: string;
  metric_value: any;
  unit: string;
  forecast_value?: any;
  intensity_index?: string;
  generation_gas?: any;
  generation_wind?: any;
  generation_solar?: any;
  generation_nuclear?: any;
  generation_biomass?: any;
  generation_hydro?: any;
  generation_imports?: any;
  raw_payload: any;
}

export interface ValidationResult {
  isValid: boolean;
  rejectionReason: string | null;
  cleanedRecord?: {
    id: string;
    source_id: string;
    timestamp: string;
    period_from: string;
    period_to: string;
    entity_id: string;
    entity_name: string;
    region: string;
    metric_name: string;
    metric_value: number;
    unit: string;
    forecast_value: number | null;
    intensity_index: string;
    generation_gas: number;
    generation_wind: number;
    generation_solar: number;
    generation_nuclear: number;
    generation_biomass: number;
    generation_hydro: number;
    generation_imports: number;
    renewable_share: number;
  };
}

export class DataCleaner {
  private existingKeys = new Set<string>();

  constructor(existingCompositeKeys: string[] = []) {
    existingCompositeKeys.forEach(k => this.existingKeys.add(k));
  }

  public validateAndClean(raw: RawInputRecord): ValidationResult {
    // 1. Missing required identifiers
    if (!raw.period_from || !raw.period_to) {
      return { isValid: false, rejectionReason: 'MISSING_REQUIRED_TIMESTAMP_FIELD' };
    }

    if (!raw.entity_id || !raw.entity_name) {
      return { isValid: false, rejectionReason: 'MISSING_REQUIRED_ENTITY_IDENTIFIER' };
    }

    // 2. Timestamp validity check
    const fromDate = new Date(raw.period_from);
    const toDate = new Date(raw.period_to);

    if (isNaN(fromDate.getTime()) || isNaN(toDate.getTime())) {
      return { isValid: false, rejectionReason: 'INVALID_TIMESTAMP_FORMAT' };
    }

    if (fromDate.getTime() >= toDate.getTime()) {
      return { isValid: false, rejectionReason: 'INVALID_INTERVAL_RANGE_START_GE_END' };
    }

    // Reject future timestamps past 24h
    const now = Date.now();
    if (fromDate.getTime() > now + 24 * 3600 * 1000) {
      return { isValid: false, rejectionReason: 'STALE_OR_FUTURE_ANOMALOUS_DATE' };
    }

    // 3. Duplicate check
    const compositeKey = `${raw.source_id}_${raw.entity_id}_${raw.period_from}`;
    if (this.existingKeys.has(compositeKey)) {
      return { isValid: false, rejectionReason: 'DUPLICATE_RECORD' };
    }

    // 4. Numeric value verification
    const val = Number(raw.metric_value);
    if (raw.metric_value === null || raw.metric_value === undefined || isNaN(val) || val < 0) {
      return { isValid: false, rejectionReason: 'INVALID_OR_NEGATIVE_NUMERIC_VALUE' };
    }

    const forecast = raw.forecast_value !== undefined && raw.forecast_value !== null ? Number(raw.forecast_value) : null;
    if (forecast !== null && (isNaN(forecast) || forecast < 0)) {
      return { isValid: false, rejectionReason: 'INVALID_FORECAST_NUMERIC_VALUE' };
    }

    // 5. Fuel generation percentages check
    const gas = Math.max(0, Number(raw.generation_gas) || 0);
    const wind = Math.max(0, Number(raw.generation_wind) || 0);
    const solar = Math.max(0, Number(raw.generation_solar) || 0);
    const nuclear = Math.max(0, Number(raw.generation_nuclear) || 0);
    const biomass = Math.max(0, Number(raw.generation_biomass) || 0);
    const hydro = Math.max(0, Number(raw.generation_hydro) || 0);
    const imports = Math.max(0, Number(raw.generation_imports) || 0);

    const sumPerc = gas + wind + solar + nuclear + biomass + hydro + imports;
    // If generation mix provided, check consistency (allowing 2.5% tolerance for rounding/other minor fuels)
    if (sumPerc > 0 && Math.abs(sumPerc - 100) > 6.0 && sumPerc > 106.0) {
      return { isValid: false, rejectionReason: 'GENERATION_PERCENTAGE_SUM_EXCEEDS_TOLERANCE' };
    }

    // Renewable share definition: wind + solar + hydro
    const renewableShare = Number((wind + solar + hydro).toFixed(1));

    // Register composite key
    this.existingKeys.add(compositeKey);

    const id = `REC_${raw.source_id}_${raw.entity_id}_${raw.period_from.replace(/[-:TZ]/g, '')}`;

    return {
      isValid: true,
      rejectionReason: null,
      cleanedRecord: {
        id,
        source_id: raw.source_id,
        timestamp: raw.period_from,
        period_from: raw.period_from,
        period_to: raw.period_to,
        entity_id: String(raw.entity_id),
        entity_name: String(raw.entity_name).trim(),
        region: String(raw.region || raw.entity_name).trim(),
        metric_name: raw.metric_name || 'Carbon Intensity',
        metric_value: Number(val.toFixed(2)),
        unit: raw.unit || 'gCO2/kWh',
        forecast_value: forecast !== null ? Number(forecast.toFixed(2)) : null,
        intensity_index: (raw.intensity_index || 'moderate').toLowerCase(),
        generation_gas: Number(gas.toFixed(1)),
        generation_wind: Number(wind.toFixed(1)),
        generation_solar: Number(solar.toFixed(1)),
        generation_nuclear: Number(nuclear.toFixed(1)),
        generation_biomass: Number(biomass.toFixed(1)),
        generation_hydro: Number(hydro.toFixed(1)),
        generation_imports: Number(imports.toFixed(1)),
        renewable_share: renewableShare
      }
    };
  }
}

File: backend/cleaner.py
"""
PulseOps Data Validation & Cleaning Engine (Python)
Evaluates field presence, ISO dates, range constraints, and fuel percentage sums.
"""
from datetime import datetime
from typing import Dict, Any, Tuple, Optional

class PythonDataCleaner:
    def __init__(self, existing_keys: Optional[set] = None):
        self.existing_keys = existing_keys or set()

    def validate_and_clean(self, raw: Dict[str, Any]) -> Tuple[bool, Optional[str], Optional[Dict[str, Any]]]:
        # 1. Missing required fields
        if not raw.get('period_from') or not raw.get('period_to'):
            return False, 'MISSING_REQUIRED_TIMESTAMP_FIELD', None

        if not raw.get('entity_id') or not raw.get('entity_name'):
            return False, 'MISSING_REQUIRED_ENTITY_IDENTIFIER', None

        # 2. Timestamp validity
        try:
            from_dt = datetime.fromisoformat(raw['period_from'].replace('Z', '+00:00'))
            to_dt = datetime.fromisoformat(raw['period_to'].replace('Z', '+00:00'))
        except Exception:
            return False, 'INVALID_TIMESTAMP_FORMAT', None

        if from_dt >= to_dt:
            return False, 'INVALID_INTERVAL_RANGE_START_GE_END', None

        # 3. Duplicate check
        composite_key = f"{raw.get('source_id')}_{raw.get('entity_id')}_{raw.get('period_from')}"
        if composite_key in self.existing_keys:
            return False, 'DUPLICATE_RECORD', None

        # 4. Numeric validation
        try:
            val = float(raw.get('metric_value'))
            if val < 0:
                return False, 'INVALID_OR_NEGATIVE_NUMERIC_VALUE', None
        except (ValueError, TypeError):
            return False, 'INVALID_OR_NEGATIVE_NUMERIC_VALUE', None

        forecast = None
        if raw.get('forecast_value') is not None:
            try:
                forecast = float(raw['forecast_value'])
                if forecast < 0:
                    return False, 'INVALID_FORECAST_NUMERIC_VALUE', None
            except (ValueError, TypeError):
                return False, 'INVALID_FORECAST_NUMERIC_VALUE', None

        # 5. Generation mix verification
        gas = max(0.0, float(raw.get('generation_gas') or 0.0))
        wind = max(0.0, float(raw.get('generation_wind') or 0.0))
        solar = max(0.0, float(raw.get('generation_solar') or 0.0))
        nuclear = max(0.0, float(raw.get('generation_nuclear') or 0.0))
        biomass = max(0.0, float(raw.get('generation_biomass') or 0.0))
        hydro = max(0.0, float(raw.get('generation_hydro') or 0.0))
        imports = max(0.0, float(raw.get('generation_imports') or 0.0))

        sum_perc = gas + wind + solar + nuclear + biomass + hydro + imports
        if sum_perc > 0 and abs(sum_perc - 100.0) > 6.0 and sum_perc > 106.0:
            return False, 'GENERATION_PERCENTAGE_SUM_EXCEEDS_TOLERANCE', None

        self.existing_keys.add(composite_key)
        cleaned_id = f"REC_{raw.get('source_id')}_{raw.get('entity_id')}_{raw['period_from'].replace(':', '').replace('-', '')}"

        cleaned = {
            'id': cleaned_id,
            'source_id': raw.get('source_id'),
            'timestamp': raw['period_from'],
            'period_from': raw['period_from'],
            'period_to': raw['period_to'],
            'entity_id': str(raw['entity_id']),
            'entity_name': str(raw['entity_name']).strip(),
            'region': str(raw.get('region') or raw['entity_name']).strip(),
            'metric_name': raw.get('metric_name', 'Carbon Intensity'),
            'metric_value': round(val, 2),
            'unit': raw.get('unit', 'gCO2/kWh'),
            'forecast_value': round(forecast, 2) if forecast is not None else None,
            'intensity_index': str(raw.get('intensity_index', 'moderate')).lower(),
            'generation_gas': round(gas, 1),
            'generation_wind': round(wind, 1),
            'generation_solar': round(solar, 1),
            'generation_nuclear': round(nuclear, 1),
            'generation_biomass': round(biomass, 1),
            'generation_hydro': round(hydro, 1),
            'generation_imports': round(imports, 1),
            'renewable_share': round(wind + solar + hydro, 1)
        }
        return True, None, cleaned
